### PR TITLE
Dev 4647.refresh.token

### PR DESCRIPTION
**Pull request by rizwansaeed**
# Pull Request Checklist

- [x ] Read the [contributing guidelines](https://github.com/finbourne/lusid-sdk-python-preview/blob/master/CONTRIBUTING.md)
- [x ] Tests pass
- [x ] Raised the PR against the `develop` branch

# Description of the PR

Add automatically refreshing access token
Refactors test for easier testing of access tokens and reuse of access tokens across tests

### CODE REVIEW

File: sdk/lusid/utilities/refreshing_token.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+from datetime import timedelta
+from collections import UserString
+
+
+class RefreshingToken(UserString):
+
+    def __init__(self, token_url, client_id, client_secret, access_token, refresh_token, expiry_offset=60):
+
+        token_data = {
+            "expires": datetime.utcnow(),
```
Comment (MikeMcGarry) [inline sdk/lusid/utilities/refreshing_token.py]: I think if you replace this with the expiry inside access_token it won't run straight away as John suggested.

File: sdk/lusid/utilities/api_client_builder.py [hunk 2]
```diff
@@ -13,62 +13,46 @@
 
 class ApiClientBuilder:
 
-    def build(self, api_configuration):
+    @staticmethod
+    def build(api_configuration, okta_response_handler = None):
```
Comment (MikeMcGarry) [inline sdk/lusid/utilities/api_client_builder.py]: What is api_configuration? Is this a string which is the path to the secrets file? Be good to have its type in the function docstring i.e. :param (string) api_configuration

File: sdk/lusid/utilities/api_configuration.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
+class ApiConfiguration:
```
Comment (MikeMcGarry) [inline sdk/lusid/utilities/api_configuration.py]: These getter and setters are cool, nice.

File: sdk/lusid/utilities/__init__.py [hunk 4]
```diff
@@ -0,0 +1,4 @@
+from lusid.utilities.api_client_builder import ApiClientBuilder
+from lusid.utilities.api_configuration import ApiConfiguration
+from lusid.utilities.api_configuration_loader import ApiConfigurationLoader
+from lusid.utilities.refreshing_token import RefreshingToken
```
Comment (MikeMcGarry) [inline sdk/lusid/utilities/__init__.py]: The ApiConfigurationLoader takes "api_configuration" as a parameter. As it returns the ApiConfiguration class I think this is a bit confusing. I'd suggest renaming this to something like "api_secrets_filename" or something to that effect and make it clear in the doc string of the load function of ApiConfigurationLoader that it is a string.

File: sdk/lusid/utilities/api_client_builder.py (hunk 2 above)
Comment (MikeMcGarry) [reply sdk/lusid/utilities/api_client_builder.py]: Repeating comment from above. The ApiConfigurationLoader takes "api_configuration" as a parameter. As it returns the ApiConfiguration class I think this is a bit confusing. I'd suggest renaming this to something like "api_secrets_filename" or something to that effect and make it clear in the doc string of the load function of ApiConfigurationLoader that it is a string.

File: sdk/lusid/utilities/api_client_builder.py [hunk 5]
```diff
@@ -13,62 +13,46 @@
...
         """
         :param api_configuration: name api configuration file
+        :param okta_response_handler: optional handler for Okta response
```
Comment (MikeMcGarry) [inline sdk/lusid/utilities/api_client_builder.py]: What type is this? Would this be a function?

File: sdk/lusid/utilities/api_client_builder.py [hunk 6]
```diff
@@ -13,62 +13,46 @@
...
-            client_id = pathname2url(client_id_raw)
-            client_secret = pathname2url(client_secret_raw)
+        # Load the configuration
+        configuration = ApiConfigurationLoader().load(api_configuration)
 
         # Prepare our authentication request
-        token_request_body = ("grant_type=password&username={0}".format(username) +
-                              "&password={0}&scope=openid client groups".format(password) +
-                              "&client_id={0}&client_secret={1}".format(client_id, client_secret))
+        token_request_body = f"grant_type=password&username={configuration.username}" \
```
Comment (MikeMcGarry) [inline sdk/lusid/utilities/api_client_builder.py]: You're going to persuade me to use f strings aren't you. Hard to see how format is better.

File: sdk/lusid/utilities/api_client_builder.py [hunk 7]
```diff
@@ -13,62 +13,46 @@
...
+
+        if okta_response_handler is not None:
+            okta_response_handler(okta_response)
 
         # Ensure that we have a 200 response code
         assert okta_response.status_code == 200
 
+        okta_json = okta_response.json()
+
         # Retrieve our api token from the authentication response
```
Comment (MikeMcGarry) [inline sdk/lusid/utilities/api_client_builder.py]: Might want to update this comment to describe what is going on.

File: sdk/lusid/utilities/api_configuration_loader.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
+import os
+import json
+from urllib.request import pathname2url
+from .api_configuration import ApiConfiguration
+
+
+class ApiConfigurationLoader:
+
+    @staticmethod
+    def load(api_configuration):
```
Comment (MikeMcGarry) [inline sdk/lusid/utilities/api_configuration_loader.py]: See earlier comment about confusion of this parameter name to the ApiConfiguration class

File: sdk/tests/test_token_refresh.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        original_token, refresh_token = self.get_okta_tokens()
+
+        refreshed_token = RefreshingToken(token_url=self.config.token_url,
+                                          client_id=self.config.client_id,
+                                          client_secret=self.config.client_secret,
+                                          access_token=original_token,
+                                          refresh_token=refresh_token)
+
+        self.assertIsNotNone(refreshed_token)
+        self.assertNotEqual(original_token, refreshed_token)
```
Comment (MikeMcGarry) [inline sdk/tests/test_token_refresh.py]: A heads up, if we don't get a refresh token straight away I imagine this second assertion will fail as the tokens should match.

File: sdk/tests/test_token_refresh.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+    def test_refresh_token_when_expired(self):
```
Comment (MikeMcGarry) [inline sdk/tests/test_token_refresh.py]: Nice way not to wait 60 minutes for the test to run!

File: sdk/tests/test_token_refresh.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+        self.assertIsNotNone(refreshed_token)
+
+        # force de-referencing the token value
+        first_value = f"{refreshed_token}"
+
+        sleep(1)
+
+        self.assertEqual(first_value, refreshed_token)
+
+    def test_can_make_header(self):
```
Comment (MikeMcGarry) [inline sdk/tests/test_token_refresh.py]: What does this test achieve?

File: sdk/lusid/utilities/api_client_builder.py (hunk 6 above)
Comment (rizwansaeed) [reply sdk/lusid/utilities/api_client_builder.py]: 👍

File: sdk/lusid/utilities/api_client_builder.py (hunk 7 above)
Comment (rizwansaeed) [reply sdk/lusid/utilities/api_client_builder.py]: done

File: sdk/lusid/utilities/__init__.py (hunk 4 above)
Comment (rizwansaeed) [reply sdk/lusid/utilities/__init__.py]: done

File: sdk/lusid/utilities/api_configuration_loader.py (hunk 8 above)
Comment (rizwansaeed) [reply sdk/lusid/utilities/api_configuration_loader.py]: done

File: sdk/tests/test_token_refresh.py (hunk 11 above)
Comment (rizwansaeed) [reply sdk/tests/test_token_refresh.py]: Initial implementation failed when used, this test was added as a regression check

File: sdk/lusid/utilities/refreshing_token.py (hunk 1 above)
Comment (rizwansaeed) [reply sdk/lusid/utilities/refreshing_token.py]: done

Comment (rizwansaeed) [reply sdk/lusid/utilities/refreshing_token.py]: fixed

**Pull request merged**